main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -28,301 +28,346 @@ import ca.uwaterloo.flix.util.Validation._
 
-    case Expression.Unary(sop, op, exp, tpe, loc) =>
+    case OccurrenceAst.Expression.Unary(sop, op, exp, tpe, loc) =>
       val e = visitExp(exp, subst0)
-      Expression.Unary(sop, op, e, tpe, loc)
+      LiftedAst.Expression.Unary(sop, op, e, tpe, loc)
 
-    case Expression.Binary(sop, op, exp1, exp2, tpe, loc) =>
+    case OccurrenceAst.Expression.Binary(sop, op, exp1, exp2, tpe, loc) =>
       val e1 = visitExp(exp1, subst0)
       val e2 = visitExp(exp2, subst0)
-      Expression.Binary(sop, op, e1, e2, tpe, loc)
+      LiftedAst.Expression.Binary(sop, op, e1, e2, tpe, loc)
 
-    case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+    case OccurrenceAst.Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
       val e1 = visitExp(exp1, subst0)
       val e2 = visitExp(exp2, subst0)
       val e3 = visitExp(exp3, subst0)
-      Expression.IfThenElse(e1, e2, e3, tpe, loc)
+      e1 match {
+        case LiftedAst.Expression.True(_) => e2
+        case LiftedAst.Expression.False(_) => e3
+        case _ => LiftedAst.Expression.IfThenElse(e1, e2, e3, tpe, loc)
+      }
 
-    case Expression.Branch(exp, branches, tpe, loc) =>
+    case OccurrenceAst.Expression.Branch(exp, branches, tpe, loc) =>
       val e = visitExp(exp, subst0)
       val bs = branches.map {
         case (sym, br) => sym -> visitExp(br, subst0)
       }
-      Expression.Branch(e, bs, tpe, loc)
+      LiftedAst.Expression.Branch(e, bs, tpe, loc)
 
-    case Expression.JumpTo(_, _, _) => exp0
+    case OccurrenceAst.Expression.JumpTo(sym, tpe, loc) => LiftedAst.Expression.JumpTo(sym, tpe, loc)
 
-    case Expression.Let(sym, exp1, exp2, occur, tpe, purity, loc) =>
-      /// Case 1:
-      /// If `exp1` occurs once and it is pure, then it is safe to inline without increasing neither the code size
-      /// nor the execution time.
-      val wantToPreInline = (occur, purity) match {
-        case (Occur.Once, Purity.Pure) => true
-        case _ => false
-      }
-      if (wantToPreInline) {
-        val subst1 = subst0 + (sym -> exp1)
-        visitExp(exp2, subst1)
+    case OccurrenceAst.Expression.Let(sym, exp1, exp2, occur, tpe, purity, loc) =>
+      /// If `sym` is never used (it is `Dead`)  and `exp1` is pure, then it is safe to remove `sym`
+      if (occur == Dead && purity == Purity.Pure) {
+        visitExp(exp2, subst0)
       } else {
-        /// Case 2:
-        /// If `e1` is trivial and pure, then it is safe to inline without increasing execution time.
-        val e1 = visitExp(exp1, subst0)
-        val wantToPostInline = (occur, purity, isTrivialExp(e1)) match {
-          case (Occur.DontInline, _, _) => false
-          case (_, Purity.Pure, true) => true
+        /// Case 1:
+        /// If `exp1` occurs once and it is pure, then it is safe to inline without increasing neither the code size
+        /// nor the execution time.
+        val wantToPreInline = (occur, purity) match {
```

Comment:
I think some named Booleans / helper functions would be useful. E.g. `isUsedOnceAndPure` ...